### PR TITLE
Added Heroku deployment instructions

### PR DESCRIPTION
**Pull request by PavanKumar06**

### CODE REVIEW

File: tutorials/heroku-deployment.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+## Step 3: Connect to GitHub for Continuous Deployment
+
+1. **Connect to GitHub Repository**
+   * In your Heroku app dashboard, scroll to the **Deployment method** section and select **GitHub**.
+   * Click **Connect to GitHub** and search for your repository.
+   * Select the repository containing your backend service code.
+
+   ![Connect to your Repo](assets/heroku-deployment/ConnectToGithub.png)
+
+2. **Configure Required Files**
```
Comment (anikeshk) [inline tutorials/heroku-deployment.md:149]: By now, most if not all teams would have added new dependencies in their `package.json`. From what I can see in the `package.json` in the assets folder is that the "start" script has been changed. So we should just ask the students to make a change in their existing `package.json`, as follows.

- Move "dotenv" from the "devDependecies" to "dependencies".
- Change the "start" script in your `server/package.json` from "ts-node server.ts" to "node dist/server.js".
- Add a new script called "start:dev" with the value "ts-node server.ts". You can use this for your local development.

Your `package.json` should look something like this:

```
{
  "dependencies": {
    ...original dependencies and any dependencies you have added
    "dotenv": "^16.4.5"
  },
  "devDependencies": {
    ...
  },
  "scripts": {
    "start": "node ./dist/server.js",
    "start:dev": "ts-node server.ts",
    "lint": "eslint .",
    "lint:fix": "eslint . --fix",
    "test": "jest -w=1 --coverage --detectOpenHandles",
    "build": "tsc",
    "start:prod": "node ./dist/server.js",
    "stryker": "stryker run"
  }
}
```

Similarly, I see only one change in the `tsconfig.json`, as follows.

- Go to `server/tsconfig.json` and scroll to the end. Modify the "include" and "exclude" values to the following:

```
"include": ["./**/*.ts"],
 "exclude": ["node_modules", "dist"]
```

If there are other similar changes we can add it as instructions. This way the students retain their existing file and dependencies.

File: tutorials/heroku-deployment.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+   * Click **Enable Automatic Deploys** to deploy your application automatically.
+
+## Step 5: Database Initialization and Environment Variables
+
+1. **Set Up Environment Variables**
+   * In your Heroku dashboard, go to **Settings -> Config Vars**
+   * Add your MongoDB connection string as a config var:
+     ```
+     Key: MONGODB_URI
+     Value: mongodb+srv://dspavankumar99:<password>@assignment.wmfzac9.mongodb.net
```
Comment (anikeshk) [inline tutorials/heroku-deployment.md]: Change the value from `mongodb+srv://dspavankumar99:<password>@assignment.wmfzac9.mongodb.net` to `<your MongoDB connection string>`

File: tutorials/heroku-deployment.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+   * Another option is to add the .env file to your server repository.
+
+2. **Run Database Initialization**
+   * Once deployed, go to **More -> Run Console** in Heroku
+
+    ![Run Console](assets/heroku-deployment/RunConsole.png)
+
+   * Run your database population command:
+     ```
+     npx ts-node populate_db.ts mongodb+srv://dspavankumar99:<password>@assignment.wmfzac9.mongodb.net/fake_so
```
Comment (anikeshk) [inline tutorials/heroku-deployment.md]: In the [MongoDB Atlas instructions](https://neu-se.github.io/CS4530-Fall-2024/activities/continuous-deployment#setup-your-mongodb-database) we have asked the students to run the `populate_db.ts` script already, so don't need to ask them to read it again - switching from Render to Heroku won't impact the MongoDB Atlas deployment.
Please remove this step.

File: tutorials/heroku-deployment.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+     ```
+   * Note: Make sure to replace `<password>` with your actual MongoDB password.
+
+   ![Run PopulateDB](assets/heroku-deployment/RunPopulateDb.png)
+
+3. **Verify Database Connection**
```
Comment (anikeshk) [inline tutorials/heroku-deployment.md]: We should have a `## Step 6: Update client env variables on Render.com` to inform students to update the server URL from the Render.com URL to the Heroku URL. These steps can be something along these lines:

1. Open the [Render Dashboard](https://dashboard.render.com/).
2. Choose the project you have created, and go back to service with type "Static Site" (your client deployment).
3. Click on the “Environment” tab.
4. Change the value of the `REACT_APP_SERVER_URL` env to the Heroku URL (you can find this in your Heroku app's "Settings" page, under the "Domains" section).
5. After the client finishes re-deploying, check if the client is sending queries to the new Heroku server deployment successfully.
6. (Optional) If the new Heroku server deployment is working correctly, delete the server deployment from Render.

File: tutorials/heroku-deployment.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
+# Steps to Set Up Heroku Deployment for Your Backend Service
```
Comment (anikeshk) [inline tutorials/heroku-deployment.md:1]: PR is created to [neu-se:all_tutorials](https://github.com/neu-se/CS4530-Fall-2024/tree/all_tutorials) but I think PRs should be created to [neu-se:main](https://github.com/neu-se/CS4530-Fall-2024/tree/main) for the Netlify auto-deployment to work.

**Review by anikeshk — CHANGES_REQUESTED** [review]

File: tutorials/heroku-deployment.md (hunk 2 above)
Comment (PavanKumar06) [reply tutorials/heroku-deployment.md]: Done

**Comment by anikeshk** [comment]
Two more things to add:

1.  [add as a note under wherever you tell the students to create a new repo for sever]

- Copy the `.gitignore` from the root of the original repository to the server repository.

2. [add as a note under wherever you tell the students to create a new repo for sever]

Once you separate the repository into one with client and one with the server, update the `.github\workflows\main.yml` in each as the following:

`server`

```
name: FakeStackOverflow CI
on: # Controls when the action will run.
  # Triggers the workflow on push or pull request events but only for the master branch. If you want to trigger the action on other branches, add here
  push:
    branches: [main]
  pull_request:
    branches: [main]

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  build-and-test: #
    # The type of runner that the job will run on
    runs-on: ubuntu-20.04

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      - name: Git Checkout
        uses: actions/checkout@v4

      - name: Set up Node.js 20.x
        uses: actions/setup-node@v4
        with:
          node-version: "20.x"

      - name: Launch MongoDB Server
        uses: supercharge/mongodb-github-action@1.11.0
        with:
          mongodb-version: "7.0"

      - name: Build and lint frontend
        if: ${{ always() }}
        env:
          REACT_APP_SERVER_URL: http://localhost:8000
        run: cd client; npm ci && npm run build && npm run lint

```

`client`

```
name: FakeStackOverflow CI
on: # Controls when the action will run.
  # Triggers the workflow on push or pull request events but only for the master branch. If you want to trigger the action on other branches, add here
  push:
    branches: [main]
  pull_request:
    branches: [main]

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  build-and-test: #
    # The type of runner that the job will run on
    runs-on: ubuntu-20.04

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      - name: Git Checkout
        uses: actions/checkout@v4

      - name: Set up Node.js 20.x
        uses: actions/setup-node@v4
        with:
          node-version: "20.x"

      - name: Launch MongoDB Server
        uses: supercharge/mongodb-github-action@1.11.0
        with:
          mongodb-version: "7.0"

      - name: Build backend service
        if: ${{ always() }}
        run: cd server; npm ci

      - name: Test backend server
        if: ${{ always() }}
        env:
          # Pass the environmental variables for the backend tests to use
          MONGODB_URI: mongodb://localhost:27017
        run: |
          cd server

          npm run start & sleep 10

          echo "Checking if the server is running..."
          curl --fail 'http://localhost:8000/question/getQuestion?order=newest&search=' || (echo "Server failed to start" && killall node && exit 1)

          echo "Server started successfully. Now stopping..."
          killall node

          npm run test

      - name: Lint backend
        if: ${{ always() }}
        run: cd server; npm run lint

```

3. [add as a note in the end]

  - In case you are still running into memory issues in the Heroku deployment, you can update the "Dynos" type.
  - Add the steps from here: https://help.heroku.com/XX3MKY2E/how-do-i-change-the-dyno-types-for-my-application 
  - Choose the type carefully as choosing a higher type will result in a higher cost, which will reduce your Platform Credits.
  - Screenshot of choosing a higher Dyno type:

image

image

File: tutorials/heroku-deployment.md (hunk 3 above)
Comment (PavanKumar06) [reply tutorials/heroku-deployment.md]: Done

File: tutorials/heroku-deployment.md (hunk 1 above)
Comment (PavanKumar06) [reply tutorials/heroku-deployment.md:149]: Done

File: tutorials/heroku-deployment.md (hunk 4 above)
Comment (PavanKumar06) [reply tutorials/heroku-deployment.md]: Done

**Comment by PavanKumar06** [comment]
I have resolved the comments made by Anikesh. If everything works fine I will close this PR and raise a new PR for Module13 Activity.md file

File: tutorials/heroku-deployment.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+1. **Create a New Repository**
+   * You must create a new Repository on Github which will contain only your server code.
+   * Crate a new Repository on Github and copy paste your server folder.
+   * Copy the .gitignore from the root of the original repository to the server repository.
+
+2. **Update Yaml file**
+   * Once the client and server folders are separated into different repositories, update the .github\workflows\main.yml in each as the following:
+
+   # Server - GitHub Actions CI Workflow Configuration
```
Comment (VatsalMehta27) [inline tutorials/heroku-deployment.md:36]: This is for client

File: tutorials/heroku-deployment.md [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+         steps:
+            - name: Git Checkout
+            uses: actions/checkout@v4
+
+            - name: Set up Node.js 20.x
+            uses: actions/setup-node@v4
+            with:
+               node-version: "20.x"
+
+            - name: Launch MongoDB Server
```
Comment (VatsalMehta27) [inline tutorials/heroku-deployment.md:66]: Not needed in client

File: tutorials/heroku-deployment.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+               mongodb-version: "7.0"
+
+            - name: Build and lint frontend
+            if: ${{ always() }}
+            env:
+               REACT_APP_SERVER_URL: http://localhost:8000
+            run: cd client; npm ci && npm run build && npm run lint
+      ```
+   
+   # Client - GitHub Actions CI Workflow Configuration
```
Comment (VatsalMehta27) [inline tutorials/heroku-deployment.md:78]: Should be server

**Pull request closed without merging**